### PR TITLE
Check for Valid Node Id, Fabric Id, and CATs Values When Processing CHIP Certs

### PR DESCRIPTION
**Pull request by emargolis**
#### Problem
  -- Node Id should be in a range [1, 0xFFFFFFEFFFFFFFFF]
  -- Fabric Id should be different from 0
  -- CAT should be different from 0

#### Change overview
Added checks to validate that values encoded in the certificates are correct.
 
#### Testing
Existing tests

### CODE REVIEW

File: src/credentials/CHIPCert.cpp [hunk 1]
```diff
@@ -677,15 +681,15 @@ CHIP_ERROR ChipDN::GetCertFabricId(uint64_t & fabricId) const
 {
     uint8_t rdnCount = RDNCount();
 
-    fabricId = UINT64_MAX;
+    fabricId = kUndefinedFabricIndex;
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCert.cpp:684]: I don't see how that makes sense.  The LHS is not a fabric _index_.

File: src/credentials/CHIPCert.cpp [hunk 2]
```diff
@@ -677,15 +681,15 @@ CHIP_ERROR ChipDN::GetCertFabricId(uint64_t & fabricId) const
...
 
     for (uint8_t i = 0; i < rdnCount; i++)
     {
         switch (rdn[i].mAttrOID)
         {
         case kOID_AttributeType_ChipFabricId:
-            // Ensure only one FabricID RDN present, since start value is UINT64_MAX, which is reserved and never seen.
-            VerifyOrReturnError(fabricId == UINT64_MAX, CHIP_ERROR_WRONG_CERT_TYPE);
+            // Ensure only one FabricID RDN present, since start value is kUndefinedFabricIndex, which is reserved and never seen.
+            VerifyOrReturnError(fabricId == kUndefinedFabricIndex, CHIP_ERROR_WRONG_CERT_TYPE);
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCert.cpp:692]: Again, this should not be using fabric indices.  Should be IsValidFabricId?

File: src/credentials/CHIPCert.cpp (hunk 1 above)
Comment (emargolis) [reply src/credentials/CHIPCert.cpp:684]: Of course! My typo. Fixed that

File: src/credentials/CHIPCert.cpp (hunk 2 above)
Comment (emargolis) [reply src/credentials/CHIPCert.cpp:692]: fixed

File: src/credentials/CHIPCertFromX509.cpp [hunk 3]
```diff
@@ -160,12 +156,26 @@ static CHIP_ERROR ConvertDistinguishedName(ASN1Reader & reader, TLVWriter & writ
                             attrOID == chip::ASN1::kOID_AttributeType_ChipICAId ||
                             attrOID == chip::ASN1::kOID_AttributeType_ChipRootId)
                         {
+                            if (attrOID == chip::ASN1::kOID_AttributeType_ChipNodeId)
+                            {
+                                VerifyOrReturnError(IsOperationalNodeId(chipAttr), CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCertFromX509.cpp:161]: Suggest replacing *ALL* `CHIP_ERROR_INVALID_ARGUMENT` related to subjects to `CHIP_ERROR_WRONG_CERT_SUBJECT`

File: src/tools/chip-cert/Cmd_GenCert.cpp [hunk 4]
```diff
@@ -73,19 +73,21 @@ const char * const gCmdOptionHelp =
...
     "\n"
-    "       Subject DN Fabric Id attribute (in hex).\n"
+    "       Subject DN Fabric Id attribute in hexadecimal format with upto 8 octets with or without '0x' prefix.\n"
+    "       The value should be different from 0.\n"
     "\n"
     "   -a, --subject-cat <hex-digits>\n"
     "\n"
-    "       Subject DN CHIP CASE Authentication Tag (in hex).\n"
+    "       Subject DN CHIP CASE Authentication Tag in hexadecimal format with upto 4 octets with or without '0x' prefix.\n"
+    "       The value should be different from 0.\n"
```
Comment (tcarmelveilleux) [inline src/tools/chip-cert/Cmd_GenCert.cpp:90]:
```suggestion
    "       The version subfield (lower 16 bits) should be different from 0.\n"
```

File: src/credentials/CHIPCert.cpp [hunk 5]
```diff
@@ -748,6 +756,10 @@ CHIP_ERROR ChipDN::DecodeFromTLV(TLVReader & reader)
             uint32_t chipAttr;
             VerifyOrReturnError(attrIsPrintableString == false, CHIP_ERROR_INVALID_TLV_TAG);
             ReturnErrorOnFailure(reader.Get(chipAttr));
+            if (attrOID == chip::ASN1::kOID_AttributeType_ChipCASEAuthenticatedTag)
+            {
+                VerifyOrReturnError(chipAttr != kUndefinedCAT, CHIP_ERROR_INVALID_ARGUMENT);
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CHIPCert.cpp:761]: Added new helper function as suggested.

File: src/credentials/CHIPCertFromX509.cpp (hunk 3 above)
Comment (emargolis) [reply src/credentials/CHIPCertFromX509.cpp:161]: Replaced `CHIP_ERROR_WRONG_CERT_SUBJECT` with `CHIP_ERROR_WRONG_CERT_DN` and using it where appropriate.

File: src/credentials/CHIPCertFromX509.cpp [hunk 6]
```diff
@@ -160,12 +156,26 @@ static CHIP_ERROR ConvertDistinguishedName(ASN1Reader & reader, TLVWriter & writ
...
                         }
                         else if (attrOID == chip::ASN1::kOID_AttributeType_ChipFabricId)
                         {
+                            VerifyOrReturnError(IsValidFabricId(chipAttr), CHIP_ERROR_INVALID_ARGUMENT);
                             fabric.SetValue(chipAttr);
                         }
+                        else if (attrOID == chip::ASN1::kOID_AttributeType_ChipCASEAuthenticatedTag)
+                        {
+                            VerifyOrReturnError(CanCastTo<uint32_t>(chipAttr), CHIP_ERROR_INVALID_ARGUMENT);
+                            VerifyOrReturnError(chipAttr != kUndefinedCAT, CHIP_ERROR_INVALID_ARGUMENT);
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CHIPCertFromX509.cpp:173]: done

File: src/tools/chip-cert/Cmd_GenCert.cpp [hunk 7]
```diff
@@ -261,9 +268,10 @@ bool HandleOption(const char * progName, OptionSet * optSet, int id, const char
         break;
 
     case 'a':
-        if (!ParseInt(arg, chip32bitAttr, 16))
+        if (!ParseInt(arg, chip32bitAttr, 16) || chip32bitAttr == chip::kUndefinedCAT)
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/tools/chip-cert/Cmd_GenCert.cpp:271]: Using `IsValidCASEAuthTag() `here

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Pull request merged**